internal/retry/retry.go

```diff
@@ -80,10 +84,10 @@ func (r *OutboundMiddleware) Call(ctx context.Context, request *transport.Reques
 	rereader, finish := ioutil.NewRereader(request.Body)
 	defer finish()
 	request.Body = rereader
-	boff := policy.backoffStrategy.Backoff()
+	boff := policy.opts.backoffStrategy.Backoff()
```

Comment:
Optional nit: We should avoid using the options struct we created for the options pattern as the actual data holder. IMO the struct should not be referenced after the function which received the options processes them. All relevant properties should be copied (transformed if necessary) to a target shape (OutboundMiddleware in this case).

This gives you more freedom for your options because your options struct could theoretically accept conflicting arguments which you resolve in the function that receives the options. For example, if you had,

    foo.Call(UseTimeout(time.Second), UseDeadline(time.Now.Plus(time.Second)))

And your options struct has both a `deadline` and a `timeout` field, you would decide which applies before copying the result into the target struct. The business logic of the target won't need to be aware of the different conflicting options and how to resolve them.

(This is a bad example because you would normally implement UseTimeout in terms of UseDeadline but you probably get my point.)